icu4c/source/i18n/messageformat2_function_registry.cpp

```diff
@@ -421,25 +423,174 @@ static FormattedPlaceholder notANumber(const FormattedPlaceholder& input) {
     return FormattedPlaceholder(input, FormattedValue(UnicodeString("NaN")));
 }
 
-static FormattedPlaceholder stringAsNumber(const number::LocalizedNumberFormatter& nf, const FormattedPlaceholder& input, UErrorCode& errorCode) {
+// Returns true if `c` is in the interval [`first`, `last`]
+static bool inRange(UChar32 c, UChar32 first, UChar32 last) {
+    U_ASSERT(first < last);
+    return c >= first && c <= last;
+}
+
+static bool isDigit(UChar32 c) { return inRange(c, 0x0030, 0x0039); }
+
+static int32_t parseDigit(UChar32 c) {
+    switch(c) {
+    case u'0': {
+        return 0;
+    }
+    case u'1': {
+        return 1;
+    }
+    case u'2': {
+        return 2;
+    }
+    case u'3': {
+        return 3;
+    }
+    case u'4': {
+        return 4;
+    }
+    case u'5': {
+        return 5;
+    }
+    case u'6': {
+        return 6;
+    }
+    case u'7': {
+        return 7;
+    }
+    case u'8': {
+        return 8;
+    }
+    case u'9': {
+        return 9;
+    }
+    default: {
+        // Should be unreachable
+        return 0;
+    }
+    }
+}
+
+static int32_t parseDigits(const UnicodeString& s, int32_t& i) {
+    int32_t result = 0;
+    while (isDigit(s[i])) {
+        result *= 10;
+        result += parseDigit(s[i]);
+        i++;
+    }
+    return result;
+}
+
+static double parseDecimalPart(const UnicodeString& s, int32_t& i) {
+    int32_t firstDigit = i;
+    double result = (double) parseDigits(s, i);
+    int32_t denominator = 1;
+    for (int32_t j = i; j > firstDigit; j--) {
+        denominator *= 10;
+    }
+    result /= denominator;
+    return result;
+}
+
+static double parseNumberLiteral(const FormattedPlaceholder& input, UErrorCode& errorCode) {
```

Comment:
Suggestions from @sffc 

1. Use function from `double-conversion.h`
2. Use function from `number_decimalquantity.h`
3. Use function from `util.h` called `parseNumber` to replace the helper methods that parse the integer and fractional portions.

Either option 1 or 2 should be sufficient to do what we need here. Option 1 gives a `double`, option 2 gives a `DecimalQuantity`. At least option 1 should be able to handle sign, decimal point, and exponent, and probably option 2. Option 3 is helpful but doesn't replace as much code as the other 2 options.

Reply:
I tried both (1) and (2), and the problem is that there are some strings that both functions accept, which are not valid number literals according to the MF2 grammar: for example, `01` (leading zero with no decimal point), `+1` (leading '+'), `.1` (decimal point with no leading 0), and `1.` (trailing decimal point).

Since these cases are relatively simple to check for, I just added a check before calling into `StringToDouble` (option 1). 

I didn't like (3) since calling `parseNumber` to parse the decimal part still means having to do math to add the two parts together, so it doesn't seem to simplify the code that much.

Let me know what you think.

Reply:
Looks great! Option 1 seemed the most likely & directly applicable.